In [ ]:
from google.colab import drive
drive.mount("/content/drive")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
%cd /content/drive/MyDrive/DeepLearning

/content/drive/MyDrive/DeepLearning


In [ ]:
def convert_from_ms( milliseconds ):
    seconds, milliseconds = divmod(milliseconds,1000)
    minutes, seconds = divmod(seconds, 60)
    hours, minutes = divmod(minutes, 60)
    days, hours = divmod(hours, 24)
    seconds = seconds + milliseconds/1000

    # result = "%d:%d:%d:%d"%(hours,minutes,seconds,milliseconds)
    result = "%d:%d:%d"%(minutes,seconds,milliseconds)
    return result

In [ ]:
#CLOVA api

import requests
import json

class ClovaSpeechClient:
  #You must get below two parameters by your own CLOVA account:
  #Console -> Clova Speech -> Service Builder
  #You might have to also use ObjectStorage
    # Clova Speech invoke URL
    invoke_url = ''
    # Clova Speech secret key
    secret = ''

    def req_url(self, url, completion, callback=None, userdata=None, forbiddens=None, boostings=None, wordAlignment=True, fullText=True, diarization=None):
        request_body = {
            'url': url,
            'language': 'ko-KR',
            'completion': completion,
            'callback': callback,
            'userdata': userdata,
            'wordAlignment': wordAlignment,
            'fullText': fullText,
            'forbiddens': forbiddens,
            'boostings': boostings,
            'diarization': diarization,
        }
        headers = {
            'Accept': 'application/json;UTF-8',
            'Content-Type': 'application/json;UTF-8',
            'X-CLOVASPEECH-API-KEY': self.secret
        }
        return requests.post(headers=headers,
                             url=self.invoke_url + '/recognizer/url',
                             data=json.dumps(request_body).encode('UTF-8'))

    def req_object_storage(self, data_key, completion, callback=None, userdata=None, forbiddens=None, boostings=None,
                           wordAlignment=True, fullText=True, diarization=None):
        request_body = {
            'dataKey': data_key,
            'language': 'ko-KR',
            'completion': completion,
            'callback': callback,
            'userdata': userdata,
            'wordAlignment': wordAlignment,
            'fullText': fullText,
            'forbiddens': forbiddens,
            'boostings': boostings,
            'diarization': diarization,
        }
        headers = {
            'Accept': 'application/json;UTF-8',
            'Content-Type': 'application/json;UTF-8',
            'X-CLOVASPEECH-API-KEY': self.secret
        }
        return requests.post(headers=headers,
                             url=self.invoke_url + '/recognizer/object-storage',
                             data=json.dumps(request_body).encode('UTF-8'))

    def req_upload(self, file, completion, callback=None, userdata=None, forbiddens=None, boostings=None,
                   wordAlignment=True, fullText=True, diarization=None):
        request_body = {
            'language': 'ko-KR',
            'completion': completion,
            'callback': callback,
            'userdata': userdata,
            'wordAlignment': wordAlignment,
            'fullText': fullText,
            'forbiddens': forbiddens,
            'boostings': boostings,
            'diarization': diarization,
        }
        headers = {
            'Accept': 'application/json;UTF-8',
            'X-CLOVASPEECH-API-KEY': self.secret
        }
        print(json.dumps(request_body, ensure_ascii=False).encode('UTF-8'))
        files = {
            'media': open(file, 'rb'),
            'params': (None, json.dumps(request_body, ensure_ascii=False).encode('UTF-8'), 'application/json')
        }
        response = requests.post(headers=headers, url=self.invoke_url + '/recognizer/upload', files=files)
        return response


if __name__ == '__main__':
    #If you have direct mp3 url, use req_url function above
    # res = ClovaSpeechClient().req_url(url='https://youtu.be/NhCoBVcXwtc?si=faZ_Qio2mRjRXirc', completion='sync')

    #Elif you have mp3 file object storage(NAVER CLOUD SERVICE), use req_object_storage function above
    # res = ClovaSpeechClient().req_object_storage(data_key='data/media.mp3', completion='sync')

    #Elif you want to use local storage, use req_upload function above
    res = ClovaSpeechClient().req_upload(file='data/media.mp4', completion='sync')
    print(res.text)

b'{"language": "ko-KR", "completion": "sync", "callback": null, "userdata": null, "wordAlignment": true, "fullText": true, "forbiddens": null, "boostings": null, "diarization": null}'
{"result":"COMPLETED","message":"Succeeded","token":"4ebbf4dc54024ec485cba1ec6619fd03","version":"ncp_v2_v2.2.3-ncp-c164738-20231016_231023-b281d5fc_v4.1.5_ko_ncp_20221227_","params":{"service":"ncp","domain":"general","lang":"ko","completion":"sync","diarization":{"enable":true,"speakerCountMin":-1,"speakerCountMax":-1},"boostings":[],"forbiddens":"","wordAlignment":true,"fullText":true,"noiseFiltering":true,"resultToObs":false,"priority":0,"userdata":{"_ncp_DomainCode":"tobigs-23-2-conference-free","_ncp_DomainId":6304,"_ncp_TaskId":15042171,"_ncp_TraceId":"9ab86efa3c1844e6bb88c86466b8affc"}},"progress":100,"keywords":{},"segments":[{"start":840,"end":4340,"text":"테스트 영상입니다. 너도 한마디만 해줘","confidence":0.9890458,"diarization":{"label":"1"},"speaker":{"label":"1","name":"A","edited":false},"words":[[1490,19

In [ ]:
#json to dict

import json

result_json = res.text
result_dict = json.loads(result_json)
print(result_dict)

{'result': 'COMPLETED', 'message': 'Succeeded', 'token': '4ebbf4dc54024ec485cba1ec6619fd03', 'version': 'ncp_v2_v2.2.3-ncp-c164738-20231016_231023-b281d5fc_v4.1.5_ko_ncp_20221227_', 'params': {'service': 'ncp', 'domain': 'general', 'lang': 'ko', 'completion': 'sync', 'diarization': {'enable': True, 'speakerCountMin': -1, 'speakerCountMax': -1}, 'boostings': [], 'forbiddens': '', 'wordAlignment': True, 'fullText': True, 'noiseFiltering': True, 'resultToObs': False, 'priority': 0, 'userdata': {'_ncp_DomainCode': 'tobigs-23-2-conference-free', '_ncp_DomainId': 6304, '_ncp_TaskId': 15042171, '_ncp_TraceId': '9ab86efa3c1844e6bb88c86466b8affc'}}, 'progress': 100, 'keywords': {}, 'segments': [{'start': 840, 'end': 4340, 'text': '테스트 영상입니다. 너도 한마디만 해줘', 'confidence': 0.9890458, 'diarization': {'label': '1'}, 'speaker': {'label': '1', 'name': 'A', 'edited': False}, 'words': [[1490, 1920, '테스트'], [2010, 2740, '영상입니다.'], [3270, 3474, '너도'], [3474, 3920, '한마디만'], [3920, 4180, '해줘']], 'textEdited':

In [ ]:
for seg in result_dict['segments']:
    print("[", convert_from_ms(seg['start']), "~", convert_from_ms(seg['end']), "]\n", seg['speaker']['name'], ":", seg['textEdited'] )
    print()

[ 0:0:840 ~ 0:4:340 ]
 A : 테스트 영상입니다. 너도 한마디만 해줘

[ 0:4:340 ~ 0:5:140 ]
 B : 테스트입니다.



In [ ]:
# papago 번역 API 사용 - 함수 활용
import requests
import json

# translate 함수 선언
def translate(text, source, target):
    #GET CLIENT_ID, CLIENT_SECRET from "NAVER DEVELOPER CENTER", INSTEAD OF CLOVA
    CLIENT_ID, CLIENT_SECRET = '', ''
    #DO NOT CHANGE URL
    url = 'https://openapi.naver.com/v1/papago/n2mt'
    headers = {
        'Content-Type': 'application/json',
        'X-Naver-Client-Id': CLIENT_ID,
        'X-Naver-Client-Secret': CLIENT_SECRET
    }
    data = {'source': source, 'target': target, 'text': text}
    response = requests.post(url, json.dumps(data), headers=headers)

    result = response.json()['message']['result']['translatedText']
    return result


# 번역할 문장 입력 후 함수에 전달
# text = input()

# parsed_data = LIST OF (TIMESTAMP, SPEAKER, TRANS)
parsed_sequence = []
for seg in result_dict['segments']:

    timestamp = (seg['start'], seg['end'])
    speaker = seg['speaker']['name']
    trans = translate(seg['textEdited'] , 'ko', 'en')
    parsed_data = (timestamp, speaker, trans)
    parsed_sequence.append(parsed_data)

print(parsed_sequence)



[((840, 4340), 'A', 'This is a test video. Please say something'), ((4340, 5140), 'B', "It's a test.")]


In [ ]:

# 입력 비디오 파일 경로

import cv2

def extract_subclip(input_video, time_segments):
    cap = cv2.VideoCapture(input_video)

    # Get the frames per second (fps) of the input video
    fps = cap.get(cv2.CAP_PROP_FPS)

    for i, (start_timestamp, end_timestamp) in enumerate(time_segments):
        start_frame = int(start_timestamp * fps / 1000)
        end_frame = int(end_timestamp * fps / 1000)

        # Set the start frame position
        cap.set(cv2.CAP_PROP_POS_FRAMES, start_frame)

        # Read frames from start to end
        frames = []
        while cap.get(cv2.CAP_PROP_POS_FRAMES) <= end_frame:
            ret, frame = cap.read()
            if not ret:
                break
            frames.append(frame)

        # Write frames to a new video file
        output_file = f"{input_video}_segment_{i}.mp4"
        out = cv2.VideoWriter(output_file, cv2.VideoWriter_fourcc(*'mp4v'), fps, (frames[0].shape[1], frames[0].shape[0]))
        for frame in frames:
            out.write(frame)
        out.release()

        print(f"Segment {i}: {start_timestamp} to {end_timestamp} saved as {output_file}")

    cap.release()

# Example usage:
input_video = "data/media.mp4"
time_segments = [parsed_data[0] for parsed_data in parsed_sequence]

extract_subclip(input_video, time_segments)

Segment 0: 840 to 4340 saved as data/media.mp4_segment_0.mp4
Segment 1: 4340 to 5140 saved as data/media.mp4_segment_1.mp4


# **실험 1**

뉴스 / 화자 1명

https://youtu.be/NhCoBVcXwtc?si=faZ_Qio2mRjRXirc

**[공식 캡션]**


전 세계 곳곳에서 기후 변화로 인한 자연재해가 잇따르는 가운데 지구 온난화가 목표치보다 빠르게 진행되고 있다는 경고음이 나왔습니다.

앞으로 5년 안에 기후 변화의 마지노선이 깨질 확률이 크게 높아진 것으로 조사됐습니다.

지난 2015년, 전 세계 197개 회원국은 파리기후변화협약을 맺었습니다.

지구 온도의 상승 폭을 산업화 이전보다 1.5도 안으로 제한하도록 노력한다는 내용이었습니다.

'지구 기온변화 1.5도'가 인류 생존의 위협을 막기 위한 최종 마지노선이라는 겁니다.

하지만 이 목표가 깨지는 데는 그리 오랜 시간이 걸리지 않을 것으로 보입니다.

세계 기상기구 WMO는 앞으로 5년 안에 지구 기온 상승 폭이 1.5도에 도달할 확률이 66%에 달한다고 밝혔는데요,

함께 들어보겠습니다.

[페테리 탈라스 / 세계기상기구 사무총장 : 앞으로 5년 동안 (지구 온도 상승 폭이) 1.5도를 넘어설 확률은 66%입니다. 라니냐 단계가 끝나면 향후 5년 동안 기록상 가장 따뜻한 해가 될 것이 거의 확실합니다.]

향후 5년간 1.5도의 기온 상승이 일어날 확률은 20% 미만이라던 3년 전 예측이 급격히 뛰어오른 겁니다.

이런 급격한 변화를 예상하는 건 지구 온난화 현상과 함께 동태평양 열대 해역 수온이 평년보다 0.5도 이상 올라가는 '엘니뇨 현상' 때문인데요,

특히 올해는 수온이 1.5도에서 2도 이상 높아지는 '슈퍼 엘니뇨' 현상이 발생할 가능성이 큰 것으로 조사됐습니다.

세계기상기구는 이 엘니뇨 현상이 기후변화와 결합해 전 세계 곳곳에 폭염과 홍수, 가뭄을 일으킬 거라 경고했는데요,

인류에 닥칠 기후 재앙을 막기 위해 노력할 수 있는 시간은 지금도 빠르게 줄어들고 있습니다.


**[CLOVA 결과]**

[ 0:0:0 ~ 0:6:360 ]
 A : 지난 2015년 전 세계 197개 회원국은 파리 기후변화 협약을 맺었습니다.

[ 0:6:360 ~ 0:20:200 ]
 A : 지구 온도의 상승폭을 산업화 이전보다 1.5도 안으로 제한하도록 노력한다는 내용이었습니다. 지구 기온 변화 1.5도가 인류 생존의 위협을 막기 위한 최종 마지노선이라는 겁니다.

[ 0:20:219 ~ 0:29:299 ]
 A : 하지만 이 목표가 깨지는 데는 그리 오랜 시간이 걸리지 않을 것으로 보입니다. 세계기상기구 wmo는 앞으로 5년 안에 지구 기온 상

[ 0:29:299 ~ 0:36:900 ]
 A : 폭이 1.5도에 도달할 확률이 66%에 달한다고 밝혔는데요. 함께 들어보겠습니다.

[ 1:3:870 ~ 1:12:920 ]
 A : 향후 5년간 1.5도의 기온 상승이 일어날 확률은 20% 미만이라던 3년 전의 예측이 급격히 뛰어오른 겁니다.

[ 1:12:920 ~ 1:23:480 ]
 A : 이런 급격한 변화를 예상하는 건 지구 온난화 현상과 함께 동태평양 열대 해역 수온이 평년보다 0.5도 이상 올라가는 엘리뇨 현상 때문인데요.

[ 1:23:480 ~ 1:31:659 ]
 A : 특히 올해는 수온이 1.5도에서 2도 이상 높아지는 슈퍼 엘리뇨 현상이 발생할 가능성이 큰 것으로 조사됐습니다.

[ 1:31:659 ~ 1:40:620 ]
 A : 세계기상기구는 이 엘리뇨 현상이 기후변화와 결합해 전 세계 곳곳에 폭염과 홍수 가뭄을 일으킬 것이라고 경고했는데요.

[ 1:40:620 ~ 1:46:660 ]
 A : 인류에 닥칠 기후재앙을 막기 위해 노력할 수 있는 시간은 지금도 빠르게 줄어들고 있습니다.

**[PAPAGO 결과]**

[ 0:0:0 ~ 0:6:360 ]
 A : Back in 2015, 197 member states around the world signed the Paris Climate Change Convention.

[ 0:6:360 ~ 0:20:200 ]
 A : It was about trying to limit the rise in global temperature to within 1.5 degrees than before industrialization. The global temperature change of 1.5 degrees is the final Maginot Line to prevent the threat of human survival.

[ 0:20:219 ~ 0:29:299 ]
 A : But it won't take long for this goal to be broken. The World Meteorological Organization (WMO) is expected to raise global temperature in the next five years

[ 0:29:299 ~ 0:36:900 ]
 A : There's a 66 percent chance that the width will reach 1.5 degrees. Let's listen to it together.

[ 1:3:870 ~ 1:12:920 ]
 A : The prediction three years ago that the probability of a 1.5-degree temperature increase over the next five years was less than 20%.

[ 1:12:920 ~ 1:23:480 ]
 A : This rapid change is expected due to global warming and the Eligno phenomenon, in which the temperature of tropical waters in the eastern Pacific Ocean rises more than 0.5 degrees above normal.

[ 1:23:480 ~ 1:31:659 ]
 A : In particular, it has been investigated that this year, there is a high possibility of a super-elieno phenomenon, in which the water temperature rises from 1.5 degrees to more than 2 degrees.

[ 1:31:659 ~ 1:40:620 ]
 A : The World Meteorological Organization warned that this Elinor phenomenon, combined with climate change, will cause heat waves and flood droughts around the world.

[ 1:40:620 ~ 1:46:660 ]
 A : The time we can try to prevent climate catastrophe that is coming to humanity is still running out fast.

# **실험 2**

영화 예고편 / 화자 다수 / 빠른 장면 전환 / 배경음악

https://youtu.be/rGR0YOtscNQ?si=0WnlNV_ITJZUFFzP

**[CLOVA 결과]**

[ 0:0:350 ~ 0:6:350 ]
 A : 소년은 꿈이 있어야 하는 법이여 나의 꿈은 말이여

[ 0:8:450 ~ 0:17:277 ]
 A : 안 맞고 사는데 해교를 안 댕기면 안 되냐 해교가 봐야 맨날 떠들고 차 맞는 게 일인데

[ 0:17:277 ~ 0:25:515 ]
 B : 식으로 빼는지 이제 다 파악했단 말이에요. 근데 이렇게 이거를 가우리 전부 다 다 진시적이야 자

[ 0:25:515 ~ 0:35:60 ]
 C : 아산 백호 어또 시방 그 그 놈이 우리 부여 원으로 전학을 온댜 십칠 대 일로 붙은 적 있는디 별명이 백호야 그 놈 이름이 뭔디

[ 0:38:560 ~ 0:40:885 ]
 D : 아 선배 고 차녀

[ 0:40:885 ~ 0:42:900 ]
 A : 제가 맞는 거 같은디

[ 0:46:30 ~ 0:46:950 ]
 D : 뭘

[ 0:46:950 ~ 0:49:0 ]
 C : 이 반사 진경 뭐야

[ 0:49:900 ~ 0:52:907 ]
 B : 말로야 왜 이러는 겨

[ 0:52:907 ~ 0:56:140 ]
 E : 백호 정병택 아 술 한 잔 하고 있네 진짜

**[PAPAGO 결과]**

[ 0:0:350 ~ 0:6:350 ]
 A : A boy should have a dream, my dream is a horse
[ 0:8:450 ~ 0:17:277 ]
 A : It's hard not to get hit by a bridge. It's hard to get hit by a car
[ 0:17:277 ~ 0:25:515 ]
 B : I know how to take it out. But we're all serious
[ 0:25:515 ~ 0:35:60 ]
 C : Ahsan Baekho, Eodoh Sibang. That guy transferred to Buyeowon, and his nickname is Baekho, what's his name
[ 0:38:560 ~ 0:40:885 ]
 D : Oh, senior and second daughter
[ 0:40:885 ~ 0:42:900 ]
 A : I think I'm right
[ 0:46:30 ~ 0:46:950 ]
 D : What
[ 0:46:950 ~ 0:49:0 ]
 C : What's with this reflective mirror
[ 0:49:900 ~ 0:52:907 ]
 B : What's wrong with you
[ 0:52:907 ~ 0:56:140 ]
 E : Baekho, Jung Byeongtaek, you're drinking

# **실험 3**

예능 / 화자 2명 / 배경음이나 효과음 있음

https://www.youtube.com/watch?v=wpMTMMt9t9I&t=12s

[ 0:0:0 ~ 0:9:7 ]
 A : 이게 촬영할 때도 약간 공대생 면모가 드러났다고 하는데 와이어 연기를 앞두고 병렬로 매야 한다 이런 얘기를 했다는데 이게 무슨 얘기예요.

[ 0:9:7 ~ 0:17:400 ]
 B : 그게 전우치 찍을 때였는데 그때 매일매일 이제 와이어를 타던 나날의 연속이었어요.

[ 0:17:400 ~ 0:28:405 ]
 B : 줄 하나 달고 가끔씩 막 아파트 3층 4층 높이 올라가면 너무 무섭잖아요. 무섭죠 무술 감독님한테 정동훈 감독님한테 물어보면 이거 줄 안 끊어져요. 이러면

[ 0:28:405 ~ 0:36:690 ]
 B : 아 절대 안 끊어지지 이거 천 키로를 견뎌 천 키로를 이거 한 번도 줄이 끊어진 적이 없어요. 있지 옛날에 말이야

[ 0:36:690 ~ 0:41:629 ]
 C : 그러니까 너무 걱정이 되잖아요. 지 되죠

[ 0:41:629 ~ 0:54:385 ]
 B : 그래서 아 이게 줄을 그래도 우리 진짜 떨어져서 팔다리 부러질 정도면 그래 한 줄만 달아도 머리 깨질 정도면 진짜 두 개 달자 그러니까

[ 0:54:385 ~ 0:57:637 ]
 B : 여기 이렇게 두 개 달았잖아 이런데 위에는 하나로 돼 있는데

[ 0:57:637 ~ 0:59:137 ]
 C : 그래요

[ 0:59:137 ~ 1:2:637 ]
 B : 여기 두 개 달면 뭐 하냐고 저기가 하난데

[ 1:3:637 ~ 1:6:137 ]
 B : 여기만 병렬로 달면 뭐 하냐 저기

[ 1:6:137 ~ 1:7:637 ]
 A : 병렬이어야 한다

[ 1:7:637 ~ 1:9:645 ]
 B : 이걸 병렬로 달아야지

[ 1:9:645 ~ 1:19:940 ]
 A : 여기 보니까 정동욱 무술 감독님께 의견을 드렸대요 에프는 엠에이니까 가속도가 지금 얼만데 천 킬로를 건진다 해도 이 정도 속도로 떨어지면은 끊어질 수 있다.

[ 1:19:940 ~ 1:30:794 ]
 A : 그래서 감독님이 그럼 두 개 매 그래서 지금 맨 건 직렬인데 분산이 안 되니까 병렬로 해야 된다 그 역시 이게 공대생답게 그랬더니

[ 1:30:794 ~ 1:33:390 ]
 A : 감독님이 뭐라고 그러셨어요.

[ 1:34:30 ~ 1:38:515 ]
 B : 한 1초 2초의 정적이 있더니

[ 1:38:515 ~ 1:41:25 ]
 C : 그냥 해

[ 1:43:480 ~ 1:49:987 ]
 D : 요즘 말로 쌉소리 하지 말고 고 그냥 쌉소리 하지 말고 해 우리가 다 해본 거야 다

[ 1:49:987 ~ 1:53:140 ]
 C : 맞아요. 맞아 딱 그런 표정이었어요.


[ 0:0:0 ~ 0:9:7 ]
 A : I heard that you looked like an engineering student during the shoot. You have to tie it in parallel before the wire act I heard you said this. What are you talking about.

[ 0:9:7 ~ 0:17:400 ]
 B : That was when I was shooting Jeon Woo-chi, and it was a series of days where I rode wires every day.

[ 0:17:400 ~ 0:28:405 ]
 B : It's so scary when you put on a rope and go up to the 3rd and 4th floors of the apartment building. It's scary. If you ask the martial arts director and director Jeong Dong-hoon, the rope won't break

[ 0:28:405 ~ 0:36:690 ]
 B : It's never cut off. The rope has never cut off. You know, a long time ago

[ 0:36:690 ~ 0:41:629 ]
 C : That's why I'm so. You can

[ 0:41:629 ~ 0:54:385 ]
 B : So, if we fall off and break our arms and legs, let's hang two strings if we break our heads

[ 0:54:385 ~ 0:57:637 ]
 B : There are two here. There's one on the top


[ 0:57:637 ~ 0:59:137 ]
 C : Yeah.

[ 0:59:137 ~ 1:2:637 ]
 B : What's the point of putting two here? There's only one over there

[ 1:3:637 ~ 1:6:137 ]
 B : What's the point of putting it in parallel

[ 1:6:137 ~ 1:7:637 ]
 A : It should be parallel

[ 1:7:637 ~ 1:9:645 ]
 B : You have to put this in parallel

[ 1:9:645 ~ 1:19:940 ]
 A : Looking at it here, he gave his opinion to martial arts director Jeong Dong-wook. F is an M. So even if the acceleration is about 1,000 kilometers, it can be cut off if it drops at this speed.

[ 1:19:940 ~ 1:30:794 ]
 A : So the director said, "I'm going to buy two. The one I'm wearing is in series, but I can't separate them. So I have to do it in parallel." He was like an engineering student

[ 1:30:794 ~ 1:33:390 ]
 A : The director said something.

[ 1:34:30 ~ 1:38:515 ]
 B : There was a second and two seconds of silence

[ 1:38:515 ~ 1:41:25 ]
 C : Just do it

[ 1:43:480 ~ 1:49:987 ]
 D : Don't talk about it. Don't talk about it. We've done it all

[ 1:49:987 ~ 1:53:140 ]
 C : That's right. That's right. That's exactly like that.